### PR TITLE
bugfix(belt): Fix a race condition

### PR DESCRIPTION
**Pull request by xaionaro**
With the unit-test, but without the fix:
```
==================
WARNING: DATA RACE
Write at 0x00c0000b8190 by goroutine 9:
  github.com/facebookincubator/go-belt.(*Belt).updateTools()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt.go:149 +0x348
  github.com/facebookincubator/go-belt.(*Belt).Tools()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt.go:130 +0xda
  github.com/facebookincubator/go-belt.TestRace.func2()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt_test.go:54 +0x74

Previous read at 0x00c0000b8190 by goroutine 8:
  github.com/facebookincubator/go-belt.(*Belt).Tools()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt.go:125 +0x44
  github.com/facebookincubator/go-belt.TestRace.func1()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt_test.go:49 +0x74

Goroutine 9 (running) created at:
  github.com/facebookincubator/go-belt.TestRace()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt_test.go:52 +0x756
  testing.tRunner()
      /home/xaionaro/.gimme/versions/go1.20.linux.amd64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /home/xaionaro/.gimme/versions/go1.20.linux.amd64/src/testing/testing.go:1629 +0x47

Goroutine 8 (running) created at:
  github.com/facebookincubator/go-belt.TestRace()
      /home/xaionaro/go/src/github.com/facebookincubator/go-belt/belt_test.go:47 +0x691
  testing.tRunner()
      /home/xaionaro/.gimme/versions/go1.20.linux.amd64/src/testing/testing.go:1576 +0x216
  testing.(*T).Run.func1()
      /home/xaionaro/.gimme/versions/go1.20.linux.amd64/src/testing/testing.go:1629 +0x47
==================
--- FAIL: TestRace (0.00s)
    testing.go:1446: race detected during execution of test
```

With the fix:
```
ok  	github.com/facebookincubator/go-belt	0.021s
```

### CODE REVIEW

File: belt.go [hunk 1]
```diff
@@ -122,8 +125,8 @@ func (belt *Belt) WithTool(toolID ToolID, tool Tool) *Belt {
 //
 // Do not modify the output of this function! It is for reading only.
 func (belt *Belt) Tools() Tools {
-	if len(belt.tools) == 0 {
-		return belt.tools
+	if belt.haveTools {
+		return nil
```
Comment (xaionaro) [inline belt.go]: The fix.

File: belt.go [hunk 2]
```diff
@@ -102,7 +104,7 @@ func (belt *Belt) Fields() field.AbstractFields {
 func (belt *Belt) WithTool(toolID ToolID, tool Tool) *Belt {
 	clone := belt.clone()
 
-	tools := make(Tools, len(clone.tools))
+	tools := make(Tools, len(clone.tools)+1)
```
Comment (xaionaro) [inline belt.go:107]: Just a small optimization, feel free to push back on this, since the change is not related to this PR.

File: belt.go [hunk 3]
```diff
@@ -122,8 +125,8 @@ func (belt *Belt) WithTool(toolID ToolID, tool Tool) *Belt {
 //
 // Do not modify the output of this function! It is for reading only.
 func (belt *Belt) Tools() Tools {
-	if len(belt.tools) == 0 {
-		return belt.tools
+	if !belt.haveTools {
+		return nil
```
Comment (xaionaro) [inline belt.go:129]: The fix.

**Comment by u1f35c** [comment]
There's not enough context in the commit message for me to understand what the race actually is and why this is an appropriate fix. It appears to be collision on read/write to the tools element, but the fix is just to switch to setting/checking a bool instead, but not an atomic.Bool?

**Comment by xaionaro** [comment]
> There's not enough context in the commit message for me to understand what the race actually is and why this is an appropriate fix. It appears to be collision on read/write to the tools element, but the fix is just to switch to setting/checking a bool instead, but not an atomic.Bool?

So there are two bits of knowledge contained in the `tools` field:
* The amount of `Tool`-s.
* Each `Tool`.

The second thing is calculated lazily and would be a source of a race, if we access it where the fix is. The first thing is pre-defined and never change for a `Belt` instance.

In the initial code I used `len(tools)` as the source of amount of `Tool`-s and it felt totally legit. But I was wrong: when I lazily update each `Tool` I then rewrite the whole `tools` field, which also rewrites the length of the slice/map (with the same value); and this is not legit, because when you rewrite a variable (even with the same value) you can theoretically get a corruption (at least as far as I understand).

So the fix is: stop using `len(tools)` and use a separate variable to store if we have any tools.

Let me know if more details are needed or if I should add a comment in the code somewhere.

**Review by u1f35c — APPROVED** [review]

**Pull request merged**